src/webgpu/capability_info.ts

```diff
@@ -100,41 +119,146 @@ export type PerStageBindingLimitType =
+  plainSamp:   { resource: 'sampler',             perStageBindingLimitType: 'sampler',         perPipelineBindingLimitType: 'sampler',         } as BindingKind,
+  compareSamp: { resource: 'comparison-sampler',  perStageBindingLimitType: 'sampler',         perPipelineBindingLimitType: 'sampler',         } as BindingKind,
+  sampledTex:  { resource: 'sampled-textureview', perStageBindingLimitType: 'sampled-texture', perPipelineBindingLimitType: 'sampled-texture', } as BindingKind,
+  storageTex:  { resource: 'storage-textureview', perStageBindingLimitType: 'storage-texture', perPipelineBindingLimitType: 'storage-texture', } as BindingKind,
+};
+
+const kValidStagesAll = {
+  validStages: C.ShaderStage.Vertex | C.ShaderStage.Fragment | C.ShaderStage.Compute,
+};
+const kValidStagesCompute = { validStages: C.ShaderStage.Compute };
+
+// Buffer bindings
+
+type BufferBindingType = 'uniform-buffer' | 'storage-buffer' | 'readonly-storage-buffer';
+export const kBufferBindingTypeInfo: {
+  readonly [k in BufferBindingType]: {
+    readonly usage: C.BufferUsage;
+  } & BindingTypeInfo;
+} = /* prettier-ignore */ {
+  'uniform-buffer':          { usage: C.BufferUsage.Uniform, ...kBindingKind.uniformBuf,  ...kValidStagesAll,     },
+  'storage-buffer':          { usage: C.BufferUsage.Storage, ...kBindingKind.storageBuf,  ...kValidStagesCompute, },
+  'readonly-storage-buffer': { usage: C.BufferUsage.Storage, ...kBindingKind.storageBuf,  ...kValidStagesAll,     },
+};
+export const kBufferBindingTypes = keysOf(kBufferBindingTypeInfo);
+
+// Sampler bindings
+
+type SamplerBindingType = 'sampler' | 'comparison-sampler';
+export const kSamplerBindingTypeInfo: {
+  readonly [k in SamplerBindingType]: {
+    // Add fields as needed
+  } & BindingTypeInfo;
+} = /* prettier-ignore */ {
+  'sampler':                   { ...kBindingKind.plainSamp,   ...kValidStagesAll,     },
+  'comparison-sampler':        { ...kBindingKind.compareSamp, ...kValidStagesAll,     },
+};
+export const kSamplerBindingTypes = keysOf(kSamplerBindingTypeInfo);
+
+// Texture bindings
+
+type TextureBindingType =
+  | 'sampled-texture'
+  | 'writeonly-storage-texture'
+  | 'readonly-storage-texture';
+export const kTextureBindingTypeInfo: {
+  readonly [k in TextureBindingType]: {
+    readonly usage: C.TextureUsage;
+  } & BindingTypeInfo;
+} = /* prettier-ignore */ {
+  'sampled-texture':           { usage: C.TextureUsage.Sampled, ...kBindingKind.sampledTex,  ...kValidStagesAll,     },
+  'writeonly-storage-texture': { usage: C.TextureUsage.Storage, ...kBindingKind.storageTex,  ...kValidStagesCompute, },
+  'readonly-storage-texture':  { usage: C.TextureUsage.Storage, ...kBindingKind.storageTex,  ...kValidStagesAll,     },
+};
+export const kTextureBindingTypes = keysOf(kTextureBindingTypeInfo);
+
+// All binding types (merged from above)
+
+interface BindingTypeInfo {
```

Comment:
nit: Include this before other type declarations that use it in an intersection type `& BindingTypeInfo`